### PR TITLE
Deprecate case-variables

### PR DESCRIPTION
**Pull request by dkorpel**
Should start as a deprecation first of course, but let's see if any buildkite use this feature first.

### CODE REVIEW

**Comment by dkorpel** [comment]
Only in tsv_utils:
```
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(530): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(543): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(560): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(569): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(611): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(614): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(252): Error: template instance `tsv_utils.csv2tsv.csv2tsv!(File, BufferedOutputRange!(File))` error instantiating
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(530): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(543): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(560): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(569): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(611): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(614): Error: run-time `case` variables are deprecated. Use if-else statements instead.
/buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/eBay-tsv-utils/csv2tsv/src/tsv_utils/csv2tsv.d(983): Error: template instance `tsv_utils.csv2tsv.csv2tsv!(ubyte[], Appender!(char[]))` error instantiating
```

File: compiler/src/dmd/statementsem.d
Comment (ibuclaw) [inline compiler/src/dmd/statementsem.d:2438]: Just a nit, specify the version when it should be turned into an error.  It's easier to think about what needs moving along when you don't have to count back 10 versions from the current release.

File: changelog/dmd.deprecate-case-variables.dd
Comment (maxhaton) [inline changelog/dmd.deprecate-case-variables.dd:8]: Evidence?

https://d.godbolt.org/z/MeG9drseb suggests otherwise. It's also quite easy to fix so it feels like quite a shoddy argument

Comment (maxhaton) [reply changelog/dmd.deprecate-case-variables.dd:8]: Also note that jump tables are not magically efficient

File: changelog/dmd.deprecate-case-variables.dd
Comment (maxhaton) [inline changelog/dmd.deprecate-case-variables.dd:7]: Explain why not just try to avoid aliasing (e.g. ban ref)

File: compiler/src/dmd/statementsem.d
Comment (maxhaton) [inline compiler/src/dmd/statementsem.d:2441]:
```suggestion
                cs.deprecation("The use of run-time variables in arguments case statements is now deprecated, use an if statement instead");
```

File: changelog/dmd.deprecate-case-variables.dd
Comment (maxhaton) [inline changelog/dmd.deprecate-case-variables.dd:6]: And the alternative doesn't have duplicates?

File: changelog/dmd.deprecate-case-variables.dd
Comment (ibuclaw) [reply changelog/dmd.deprecate-case-variables.dd:8]: Technically the truth.  The reality is that smart compilers just see through it.

https://d.godbolt.org/z/M3xsqxrj8

File: changelog/dmd.deprecate-case-variables.dd
Comment (dkorpel) [reply changelog/dmd.deprecate-case-variables.dd:6]: Yes, it's checked at compile time.

File: changelog/dmd.deprecate-case-variables.dd
Comment (dkorpel) [reply changelog/dmd.deprecate-case-variables.dd:7]: The gist of it is that, currently, I see run-time case variables as a half-baked extension to the classic switch statement. We can commit going forward and fix the issues, but for the sake of simplicity, I propose we go back, since the future is not commonly used as far as I know.

File: changelog/dmd.deprecate-case-variables.dd
Comment (maxhaton) [reply changelog/dmd.deprecate-case-variables.dd:6]:
```
int x;
if (x == 3) {

}

// ...
if (x == 3) {


}
```

File: changelog/dmd.deprecate-case-variables.dd
Comment (maxhaton) [reply changelog/dmd.deprecate-case-variables.dd:7]: On the one hand it's simpler in a sense that you have less choice, but on the other hand you now actually have more rules - it's not just an (say) AssignExpression any-more, it's now AssignExpression (+ rules).

Although you're right in that it's not a super-common thing to do, the one time I have used such a pattern profitably is as an intermediary step between two styles: Allowing runtime constants is a useful bit of plasticity IMO. 

Consider what happens if you have a large switch statement, and it turns out you were too clever and the design actually doesn't fit in a C-style (const-only) switch statement. We'd be encouraging people to write in a C++ style like
```
const specialCase = calc();
if (x == specialCase) {

}
switch(x) {

}
```
rather than just writing
```
const specialCase = calc();
switch(x) {
   case specialCase:
      return 6.6260701;
}
```

I am partial to the argument about duplication particularly in the case of `final` switch. You can try to prove such properties hold, but not an easy thing to do in dmd...

File: changelog/dmd.deprecate-case-variables.dd
Comment (dkorpel) [reply changelog/dmd.deprecate-case-variables.dd:6]: I thought 'alternative' was referring to compile-time case statements.

> And the alternative doesn't have duplicates?

It can have duplicates, but there's no expectation that consecutive if-statements are mutually exclusive.

File: changelog/dmd.deprecate-case-variables.dd
Comment (dkorpel) [reply changelog/dmd.deprecate-case-variables.dd:7]: > but on the other hand you now actually have more rules - it's not just an (say) AssignExpression any-more, it's now AssignExpression (+ rules).

Currently, these case types are supported:

| Case variable           | int     | string   | other types    |
|-------------------------|---------|----------|----------------|
| enum                    | yes     | yes      | no             |
| const var               | yes     | no       | no             |
| run-time expr / mut var | no      | no       | no             |

It's definitely not 'any AssignExpression'. After deprecating case variables:

| Case variable           | int     | string   | other types    |
|-------------------------|---------|----------|----------------|
| enum                    | yes     | yes      | no             |
| run-time expr   | no      | no       | no             |

**Review by RazvanN7 — APPROVED** [review]

**Comment by dkorpel** [comment]
@RazvanN7 I think this needs approval of @WalterBright and @atilaneves

**Comment by RazvanN7** [comment]
I see it as being fairly obvious in the direction of language simplification, but anyway, you are right.

**Comment by atilaneves** [comment]
I defer to @WalterBright

File: changelog/dmd.deprecate-case-variables.dd
Comment (maxhaton) [reply changelog/dmd.deprecate-case-variables.dd:8]: i.e. not the truth, no?

A true jump table is not great for your cache so you don't even want one most of the time.

**Comment by yebblies** [comment]
https://issues.dlang.org/show_bug.cgi?id=5862 ?
#2887 ?

**Comment by RazvanN7** [comment]
I can only imagine how frustrating this is for you @yebblies .

On a different note: nice to see you again!

**Comment by RazvanN7** [comment]
ping @WalterBright

File: changelog/dmd.deprecate-case-variables.dd
Comment (ntrel) [inline changelog/dmd.deprecate-case-variables.dd:31]: It's more complicated than this to get true equivalence, x may be an expression with side effects.

**Comment by WalterBright** [comment]
I don't recall why this feature was ever implemented. Andrei had a relevant comment about it:

https://github.com/dlang/dmd/pull/2887#issuecomment-37744587

People have complained a _lot_ about D gratuitously breaking things. We need an awful good reason to break things, and this doesn't quite rise to that standard. It doesn't cause harm. If people don't like it, don't use it. (We can "undocument" it, though. I doubt many even know it exists.) Switch has a number of other things wrong with it that were copied from C.

The real problem with `switch` is much more fundamental - it isn't a "match" statement. I've propose a match statement on the n.g., and it's where D should be going. Switch can be left behind as a legacy feature.

**Comment by dkorpel** [comment]
> People have complained a *lot* about D gratuitously breaking things. We need an awful good reason to break things, and this doesn't quite rise to that standard

This is at odds with what the official vision document says:

> Identify features that are underused, or that have proven to provide little benefit in the general case. The removal of these features may be controversial, but once identified we can list them in an Appendix of the spec as features that should be avoided in new code. Such features are candidates for future deprecation and removal.

https://github.com/dlang/vision-document#simplification

Do you not agree with the vision anymore?

**Comment by maxhaton** [comment]
> > People have complained a _lot_ about D gratuitously breaking things. We need an awful good reason to break things, and this doesn't quite rise to that standard
> 
> This is at odds with what the official vision document says:
> 
> > Identify features that are underused, or that have proven to provide little benefit in the general case. The removal of these features may be controversial, but once identified we can list them in an Appendix of the spec as features that should be avoided in new code. Such features are candidates for future deprecation and removal.
> 
> https://github.com/dlang/vision-document#simplification
> 
> Do you not agree with the vision anymore?

It might be at odd with what the vision is actually trying to achieve. Removing bits and pieces here and there won't make anything particularly easier from an implementation perspective (especially not at the AST level, but there are some things lexically that we don't need e.g. D has way too many types of string literal). The big thing that will bring greater progress is reducing the exponential complexity we gave ourselves with all the mixing and matching if subsets we made possible

**Comment by dkorpel** [comment]
> Removing bits and pieces here and there won't make anything particularly easier from an implementation perspective

Yes it does. When I wanted to replace duplicate switch case checking from the naive `O(n^2)` scan to use `sort()` on cases, I found out that cases can also be variables, making things more difficult. And this isn't the first time I discover obscure language complexity by trying to implement something ostensibly simple. When I improved the `-main` switch I found out there are actually 4 (!) different entry point functions dmd needs to know about, which made things harder.

**Comment by dkorpel** [comment]
> The big thing that will bring greater progress is reducing the exponential complexity we gave ourselves with all the mixing and matching if subsets we made possible

If we can't even remove a small thing, how is removing a big thing going to fare?

**Comment by WalterBright** [comment]
@dkorpel believe me I'm sympathetic to your point of view. I've tried to remove several features in favor of simpler alternatives, such as lazy functions and the struct initializer syntax. Strong and surprising pushback resulted. Even for small changes,  there is strong feedback about "no breaking changes" and a desire for stability. People just don't like a constant drip, drip, drip of breaking changes. I'm sympathetic to that, too.

This one also comes with it being non-trivial for users to fix it.

There aren't any easy answers. If I was doing a do-over, this feature wouldn't be there. Perhaps we can simply "undocument" it for now at lease.

P.S. The optimization issues are irrelevant, because the compiler can (and does) look at the pattern of case expressions to decide how to handle it.

**Comment by WalterBright** [comment]
If you can get consensus on this in the n.g. and Symmetry, I'll approve this.

**Comment by WalterBright** [comment]
Given all the current issues people have with us removing existing language features thereby breaking random projects, I'm very sorry to say I cannot see us moving forward with this.

**Pull request closed without merging**
